Bottom sheet scrolling

dnfield opened the pull request:
This is still a bit rough, but wanted to start getting some eyes on it.

Fixes #497 

Here's what I know it still needs

- [x] Tests! Existing tests need to be updated
- [x] Write new tests for some of the new functionality
- [x] ~~Persistent sheets (e.g. setting the `Scaffold.bottomSheet`) need some work.  Right now they're getting dismissed if another sheet gets dismissed.~~ This is resolved by tightening up an existing restriction to ensure that people can't show Standard Bottom Sheets if a persistent Bottom Sheet is already visible (and vice versa).  Existing code had it covered the other way around, this just extends that.  I suspect someone out there might want to do this, but I'm not sure it makes much sense design wise and it makes the code a lot more complicated.
- [x] Settle on how to handle when a child shouldn't scroll to the top of the screen.  ~~Right now, a developer has to pass a few parameters (set `initialTop` to something valid, and set `clampTop` to true).  This is the best I've been able to come up with so far - any attempts to automate that for developers result in some kind of bad result somewhere else.~~ I think the refactor to an `initialHeightPercentage` solves a lot of this.
- [x] Determine whether everything I'm breaking absolutely needs to be broken.  I'm changing some parameters and behavior here, because the existing widget is pretty badly broken in a few ways.
- [x] Make sure this all works with a11y

----

Ways existing widget is broken:

- Can't go full screen
- Can't be dragged if it has a list/scrolling content/gesturedetector in its child
- Doesn't hide the FAB if it's given a tall height
- Doesn't differentiate very well between "Standard Bottom Sheets" and "Persistent Bottom Sheets"

This is still WIP, but I think enough is done that it's worth getting some review/feedback.

dnfield committed:
start bottom sheet refactoring

dnfield committed:
Teach ScaffoldState to hide and show FAB

dnfield committed:
move stack to proper location

dnfield committed:
Make sure to dispose animationController, add test

dnfield committed:
fix indentation, remove unused call

dnfield committed:
start work on adding to theme

dnfield committed:
Start moving controllers up to scaffold

dnfield committed:
revert unneeded changes to theme

dnfield committed:
cleanup

dnfield committed:
test for interaction between show/hide and move

dnfield committed:
add onPressed to tests

dnfield committed:
merge hide_fab

dnfield committed:
Expose value instead of controller

dnfield committed:
Merge branch 'hide_fab' into bottom_sheet

dnfield committed:
Merge branch 'master' into hide_fab

dnfield committed:
scrollTopThenContent widget

dnfield committed:
merge

dnfield committed:
remove new

dnfield committed:
typo

dnfield committed:
typo

dnfield committed:
remove new

dnfield committed:
merge with hide_fab changes

dnfield committed:
Use positioned

dnfield committed:
Scrolling mechanics and FAB hiding

dnfield committed:
Dismissed sheets, scrim

dnfield committed:
Scrim touchup, formatting

dnfield committed:
merge

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
rename, minor improvements

dnfield committed:
cleanup, propagate initialtop

dnfield committed:
Apply to modal bottom sheet

dnfield committed:
Partial support for persistent bottom sheet

dnfield commented:
Here's a playground for it: https://gist.github.com/dnfield/3a524bbbd4ebf4652c3379008469d096

dnfield committed:
prevent creating a standard bottom sheet with a persistent bottom sheet

dnfield committed:
fixes for/from tests

dnfield committed:
Merge branch 'bottom_sheet' of github.com:dnfield/flutter into bottom_sheet

dnfield committed:
fixes for/from tests

dnfield committed:
Merge branch 'master' into bottom_sheet

dnfield committed:
Fixes from tests/all tests working

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1294,22 +1388,58 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
   ///  * <https://material.google.com/components/bottom-sheets.html#bottom-sheets-persistent-bottom-sheets>
-  PersistentBottomSheetController<T> showBottomSheet<T>(WidgetBuilder builder) {
+  StandardBottomSheetController<T> showBottomSheet<T>(
+      WidgetBuilder builder, {
```
[per our discussion] It would be nice if initialTop was expressed as a percentage of the screen's height.

dnfield replied:
This is done in 2af5a78

HansMuller reviewed (commented):
Just some preliminary comments, mostly just nit picking based on reading through the code.

I think this is on the right track: making the bottom sheet's scroll controller drive the bottom sheet to the top of the screen (or off the bottom) sounds good to me.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -79,71 +74,54 @@ class BottomSheet extends StatefulWidget {
+    super.initState();
+    widget.scrollController.addTopListener(() {
+      if (widget.scrollController.top == widget.scrollController.maxTop) {
+        widget.onClosing.call();
```
Actually, should just be `widget.onClosing()` - it's guaranteed not to be null.  I think this was left over from before I was asserting that.

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
nits/update tests for MediaQuery

dnfield committed:
Refactor MaxTop, use percentage for initial height instead of top

dnfield committed:
fix recursive import

dnfield commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -288,6 +299,8 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
   final double floatingActionButtonMoveAnimationProgress;
   final FloatingActionButtonAnimator floatingActionButtonMotionAnimator;
 
+  final double bottomSheetTop;
```
bottomSheetTop, here, will acutally be the distance from the app bar to the top of the bottom sheet.  Modal bottom sheets can go past the app bar, but standard ones can't (in this implementation).  I'm not entirely sure what the spec is actually directing us to do there though.

Hixie commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -293,7 +293,7 @@ class LocalHistoryEntry {
   /// Remove this entry from the history of its associated [LocalHistoryRoute].
   void remove() {
-    _owner.removeLocalHistoryEntry(this);
+    _owner?.removeLocalHistoryEntry(this);
```
do you have a test for what led to this change?

dnfield replied:
Unfortunately I don't know why I did this, and nothing appears broken without it.  Reverting that part.

peerwaya commented:
@dnfield great work, any update on this? Looks like it's almost complete.

dnfield commented:
@peerwaya thanks!

This still has a few problems - it needs to be made more accessibility friendly, and had some bug with the persistent bottom sheet implementation that has to be resolved.

I do plan to get back to this, but it needs some more work to really be ready.

ThinkDigitalRepair commented:
> Here's a playground for it: https://gist.github.com/dnfield/3a524bbbd4ebf4652c3379008469d096

Is there a way we can test this? We don't have access to the referenced file, `bottom_material.dart`

ThinkDigitalRepair commented:
Where does the scaffold, or whatever controls it, set the size of the BottomSheet?

dnfield commented:
Should be initialHeightPercentage or something like that.

ThinkDigitalRepair commented:
I'm interested in getting this widget with custom heights, including full
screen

On Tue, Nov 27, 2018, 8:36 PM Dan Field <notifications@github.com wrote:

> Should be initialHeightPercentage or something like that.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/21896#issuecomment-442316670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV-HfWeVu37OfH-M8FTqJinNmz4EOGSTks5uzhLAgaJpZM4WqFpJ>
> .
>

sh0umik commented:
any updates on the merge ?

dnfield committed:
merge

dnfield committed:
merge

dnfield committed:
merge

dnfield committed:
merge-o

dnfield committed:
analyzer nit

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield commented:
I still have to update the tests that are now failing.

One thing I'm not clear on is how A11y should behave here.  It looks like it's almost working, but how should an A11y user be able to get the sheet "dragged back down" if it takes over full screen? A sighted user would simply drag the sheet down.  A vision impaired user using TalkBack or Voiceover doesn't seem to have a way to get out.  This is probably a problem in our existing bottom sheet as well for non-persistant ones.  How does a vision-impaired user "swipe a bottom sheet down" off the screen?

dnfield committed:
remove ?

dnfield committed:
fix localhistory logic

dnfield commented:
Latest thought for A11y on persistent bottom sheets - make them work like non-persistent bottom sheets if they end up scrolling above their initial position.  That way, the back button will work and TalkBack will give a dialog notice.

layou1989 commented:
can the bottom sheet transparent? that i have a easy way to have a border-radius bottom sheet .

dnfield commented:
@layou1989 I'm not sure I understand - can you give an example of what you're looking for?

Either way it sounds like that might be a separate issue from what this PR is trying to address.

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
Make persistent bottom sheets A11y friendly

dnfield commented:
Sorry again for the long wait - I'm to the point where I think I just need to add some tests for this and it'll be ready for review.

ThinkDigitalRepair commented:
Yeah, thank you, Dan!

On Mon, Feb 4, 2019, 3:40 PM Dan Field <notifications@github.com> wrote:

> Sorry again for the long wait - I'm to the point where I think I just need
> to add some tests for this and it'll be ready for review.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/21896#issuecomment-460457665>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV-Hfa6xr86fXyHoYCHZVksRsvTZ-Nmrks5vKMUIgaJpZM4WqFpJ>
> .
>

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
fix bugs, undo unnecessary renames

dnfield committed:
tests

dnfield commented:
@HansMuller @goderbauer this is ready for review.

dnfield commented:
After discussing a bit more with Hans and goderbauer, I believe I can make this be non-breaking (default to old behavior, opt in for new behavior).  Stay tuned...

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

layou1989 commented:
can also provider the ability when the bottom scroll more than half it close,less than half it go to top,just like old

dnfield committed:
merge

dnfield committed:
Make it opt-in

dnfield commented:
This is no longer an API break.  I've reverted the tests I updated (and/or forked them into specific tests for the new functionality).

The new API adds an explicit boolean parameter to tell the bottom sheet whether it should expect to be controlled by a `BottomSheetScrollController` (in which case there needs to be a `ScrollView` child with `primary: true`), or using the legacy method with a GestureDetector/AnimationController.

@HansMuller @goderbauer this is ready for review again.

IhorKlimov commented:
@dnfield PersistentBottomSheetController#close() doesn't work for me both when __isScrollControlled__  set to true or false

dnfield commented:
@IhorKlimov - can you share an sample with what you're trying to do that isn't working?

layou1989 commented:
@dnfield can also provider the ability when the bottom scroll more than half it close,less than half it go to top,just like old

IhorKlimov commented:
@dnfield Here it is https://gist.github.com/IhorKlimov/908be06e2229ac9f4c0e4e0edc23e59b

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
fix close when isPersistent=true, add test

dnfield commented:
@IhorKlimov thanks for the small repro! I've added a test for this and it should be working now.

dnfield committed:
Merge branch 'master' into bottom_sheet

maciejczechowski commented:
@dnfield clampTop is not working for me if using isScrollControlled = true. The BottomSheet can still be swiped down, regardless of how I set the clampTop property.
Looking at showBottomSheet method, it always passes false as persistent to _buildBottomSheet. Changing it to true disables swipe-to-dismiss, but again regardless to clampTop setting ;)

dnfield commented:
If you want a bottom sheet that can't be dismissed, you need to use the bottomSheet property on Scaffold

maciejczechowski commented:
But this is then inconsistent with added documentation:
  /// the sheet should take at the start.  The `clampTop` parameter specifies
  /// whether to force the bottom sheet to always be at least that height, and
  /// setting it to true will disable swipe down dismissal of the bottom sheet.

dnfield commented:
Ahhhh ok.  I'll take a look.

dnfield commented:
I think this is a mistake in the doc.  In fact, at this point I think I should just get rid of the `clampTop` parameter entirely - it was there before I had a better way to control whether the sheet was going to be persistent or not.

The problem is that once you introduce a non-dismissable bottom sheet, you really should not be introducing any more bottom sheets unless they're modal.  It causes weird problems with trying to figure out which sheets to leave open or close or to show hte user.  The right way to show a persistent bottom sheet is to set it as a property on the Scaffold, where we can manage things better.

What do you think of that idea?

dnfield committed:
remove clampTop

dnfield committed:
Merge branch 'bottom_sheet' of github.com:dnfield/flutter into bottom_sheet

maciejczechowski commented:
That actually makes sense  - and using bottomSheet property it works just fine with ListView :)

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+class BottomSheetScrollPosition extends ScrollPositionWithSingleContext {
+  /// Creates a new [BottomSheetScrollPosition].
+  ///
+  /// The [top], [notifier], and [minTop] parameters must not be null.  If [maxTop]
```
Also context, animateIn. And minTop, maxTop must be positive numbers.

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+    }
+  }
+
+  /// Whether the [top] will be animated initially.
```
It's not entirely clear what this means.

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+      ) {
+    _topAnimationController = AnimationController(
+      value: 1.0,
+      upperBound: 1.0,
```
1.0 is the default

HansMuller commented on packages/flutter/lib/src/material/debug.dart:
```diff
@@ -127,3 +128,36 @@ bool debugCheckHasMaterialLocalizations(BuildContext context) {
+      final Element element = context;
+      throw FlutterError(
+          'No Scaffold widget found.\n'
+              '${context.widget.runtimeType} widgets require a Scaffold widget ancestor.\n'
```
no indent here

HansMuller reviewed (commented):
Mostly superficial feedback. I haven't been through the entire PR yet.

Two non-trivial issues that came up:
- Not confident that using the screen's size to limit the bottom sheet's position is right. Might want to use the size of the enclosing route.
- How much of the BottomSheetController's API do we really need to make public?

Will keep working through this but I thought it would be good or provide a progress report :-)

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -121,31 +164,64 @@ class _BottomSheetState extends State<BottomSheet> {
     if (details.velocity.pixelsPerSecond.dy > _kMinFlingVelocity) {
       final double flingVelocity = -details.velocity.pixelsPerSecond.dy / _childHeight;
-      if (widget.animationController.value > 0.0)
+      if (widget.animationController.value > 0.0) {
```
This is existing API - I'm really not sure but if we should I think it should be added in a separate PR.

HansMuller replied:
Leaving the existing semantics as they are is an important goal. I agree that it's best to leave this alone, at least for now.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -121,31 +164,64 @@ class _BottomSheetState extends State<BottomSheet> {
+      return PrimaryScrollController(
+        controller: widget.scrollController,
+        child: Material(
+          elevation: widget.elevation,
```
We could probably use a container instead but I think that'd make tests confusing and might be a breaking change - the existing API works this way if I'm reading it correctly.

HansMuller replied:
Material widgets can be expensive however leaving things as they are (to the extent that's possible) is more important than micro optimizations. Maybe in a separate PR...

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+    assert(debugCheckHasMediaQuery(context));
+    assert(forFullScreen || debugCheckHasScaffold(context));
+
+    final double screenHeight = MediaQuery.of(context).size.height;
```
That's a good point - will `Navigator.of(context).context.size.height` work?

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+    double extraAppBarHeight = 0.0;
+    if (!forFullScreen) {
+      // Scaffold.of(context) won't work if the context is the Scaffold itself.
+      final ScaffoldState scaffold = context is StatefulElement && context.state is ScaffoldState
```
This code would find the closest scaffold given the context - but it's possible that the context is itself the scaffold state.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+
+  BottomSheetScrollPosition _position;
+
+  /// The position at which the top of the controlled widget should be displayed.
```
I changed it to "bottom sheet".  At one point when writing this I hoped to make a more abstract, reusable controller.  As I iterated it became clearer to me that I probably don't understand all the use cases and what would or wouldn't break for a bottom sheet vs some other use case.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+    // If the BottomSheet's child doesn't have a Scrollable widget in it that
+    // inherits our PrimaryScrollController, it will never become visible.
+    assert(() {
+      SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
```
The problem I'm trying to solve is that if you use this scroll controller on a bottom sheet like this:

```
BottomSheet(
  child: ListView(children: ...), // not setting primary: true
  ...
),
```

it won't animate onto the screen.  But you're right, this approach would mean if someone created the BottomSheetScrollController ahead of time for some reason and later used it on a BottomSheet a few frames latter, this assert would fire.  I suppose it could be turned into a warning instead of an assert in that case, but I'm not sure of the best way to handle that and still not confuse a developer who can't see their bottom sheet.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+  /// The [AnimationStatus] of the [AnimationController] for the [top].
+  AnimationStatus get animationStatus => _position?._topAnimationController?.status;
+
+  /// Animate the [top] value to [maxTop] for non-persistent bottom sheets, or
```
maxTop - it's a little confusing, but I named this so that `minTop` would be closer to 0 and `maxTop` would always be > `minTop`.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+
+  // NotificationCallback handling for top.
+
+  List<VoidCallback> _topListeners = <VoidCallback>[];
```
The problem is we're already a ChangeListener because we're a ScrollController.  We need to track this separately.

But I made it final etc.

HansMuller replied:
Sorry, what I meant was that may you could just use an instance of ChangeNotifier here.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+  ///
+  /// This method must not be called after [dispose] has been called.
+  ///
+  /// If a listener had been added twice, and is removed once during an
```
This was copied from `ChangeNotifier`'s class doc :)  I suppose I could make it a template instead?

dnfield commented on packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart:
```diff
@@ -0,0 +1,478 @@
+    }
+  }
+
+  bool _disposed = false;
```
I had a case where dispose was getting called multiple times - it seems like in Flutter we have a preference to treat that as an error, whereas I was treating it as safe to call multiple times.

I can't remember exactly what it broke, but if we'd like to make it illegal to call dispose multiple times I can try to fix it.

dnfield replied:
I'm able to remove this without failing any tests or breaking the functionality - taking it out.  Perhaps I needed it at some point but I don't anymore.

dnfield committed:
review

dnfield commented:
@HansMuller - Everything that's public in `BottomSheetScrollController` is used somewhere else in the framework - I don't think anything that's currently public could be made private easily.  As I mentioned in another response, my original goal with it was to create a more generic controller that could be used for other widgets, but I lost confidence as I worked along that I understood other use cases enough to keep it generic and still useful for bottom sheets.  

I'll have to investigate using the route height.

dnfield committed:
remove _disposed

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1013,6 +1036,14 @@ class Scaffold extends StatefulWidget {
   ///  * [showModalBottomSheet], which displays a modal bottom sheet.
   final Widget bottomSheet;
 
+  /// Whether the [bottomSheet] is a scrollable widget with [ScrollView.primary]
```
I believe this means that the [bottomSheet] has a Scrollable descendant (like a ListView) for which primary is true.

dnfield replied:
Has is probably better, but it also could _be_ one.  We have no way to enforce that.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
 
-  PersistentBottomSheetController<T> _buildBottomSheet<T>(WidgetBuilder builder, AnimationController controller, bool isLocalHistoryEntry) {
+  PersistentBottomSheetController<T> _buildBottomSheet<T>(
+      WidgetBuilder builder,
```
two space indent

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -2001,85 +2232,119 @@ class ScaffoldFeatureController<T extends Widget, U> {
-class PersistentBottomSheetController<T> extends ScaffoldFeatureController<_PersistentBottomSheet, T> {
+///
+/// This controller is used to display both standard and persistent bottom
+/// sheets. A bottom sheet is only persistent if it is set as
```
set as ?

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+        minTop: isPersistent ? initialHeight : 0.0,
+        isPersistent: isPersistent,
+        context: context,
+      )..addTopListener(() => setState(() {
```
We (mostly) only use `=>` for one-liners that return a value

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+        if (_bottomSheetScrollController.top < 2) {
+          floatingActionButtonVisibilityValue = 0.0;
+        } else {
+          final double screenHeight = MediaQuery.of(context).size.height;
```
Shouldn't depend on screenHeight...

dnfield replied:
I'm not sure how to resolve this - maybe we can talk more tomorrow.  The problem is finding the correct height to use during build.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+        // If we're very close to the top, just set the scale to 0.
+        // This avoids issues where a fling can send us up so fast we don't acutally
+        // fully dismiss the fab.
+        if (_bottomSheetScrollController.top < 2) {
```
It's not clear that 2 is the ideal threshold here.  Maybe we need to defend against this case in a different way?

dnfield replied:
Agreed - again, not sure of a better way to handle it, maybe we can discuss.

HansMuller commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
   /// Returns a controller that can be used to close and otherwise manipulate the
   /// bottom sheet.
   ///
+  /// The `isScrollControlled` parameter specifies whether this is a route for
```
What does "this is a route" refer to here?

dnfield replied:
Copy paste error.

HansMuller reviewed (commented):
Just another progress report.

dnfield committed:
Simplify top listener logic

dnfield committed:
formatting

dnfield committed:
merge

dnfield committed:
move the postframe callback to a more sensible location

dnfield commented:
Mainly writing this down now so I don't forget - there are three "larger" items from this review so far I'm hoping to get greater clarity on:

1. I'm not sure how to avoid using the screen height.  I agree route height would be better, but we can't access that or even the constraints at build time, which is when we need it. I think the Scaffold layout delegate could _maybe_ expose it back to us, but I'm worried about that getting very messy.
2. Right now I'm using 2 pixels as "close enough" to just treat things like we've reached the top as far as hiding the FAB goes.  I suspect there's a better way to handle this, but I'm not quite sure what.
3. I moved the PostFrameCallback to a (hopefully) more sensible location (didChangeDependencies on BottomSheetState) - however, I can't come up with a good and safe way to visit the children to see if we're ok instead of just asserting the scroll controller got a client after a frame.

dnfield commented:
I've split out a lot of the bottom sheet widget specific logic into another PR (#30058) - once we can get squared away on that I'll merge it into this, which should simplify this PR a bit.

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
all but closing

dnfield committed:
cleanup docs, remove unintended or unnecessary changes

dnfield committed:
unsaved file

dnfield committed:
closing logic

dnfield committed:
Close, make persistent sheet accessible again

dnfield commented:
Just some notes from offline discussion - going to try to make the downards notification API surface a bit smaller by exposing a smaller, simpler widget to use (that will likely just use the new ones as private API).

dnfield committed:
Merge remote-tracking branch 'upstream/master' into bottom_sheet

dnfield committed:
refactor notifier, improve test

dnfield committed:
const

rami-a commented:
Hey @dnfield I'm currently working on some overlapping BottomSheet changes in https://github.com/flutter/flutter/pull/31318

I just wanted you to be aware of that. Looks like we both had a similar idea (and implementation) regarding exposing color.

HansMuller reviewed (commented):
I apologize if a few of my comments are out of date; I started this review a week ago and just resumed now.

I've looked through everything except the tests

dnfield commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -312,6 +364,11 @@ class _ModalBottomSheetRoute<T> extends PopupRoute<T> {
 /// sheet is closed.
 ///
+/// The `isScrollControlled` parameter specifies whether this is a route for
+/// a bottom sheet that will utilize [DraggableScrollableSheet]. If you wish
```
I've updated to clarify that if you wish to have this and have it draggable.  Otherwise the scrollable will scroll but the sheet won't grow.

dnfield commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -325,18 +382,26 @@ class _ModalBottomSheetRoute<T> extends PopupRoute<T> {
   @required BuildContext context,
   @required WidgetBuilder builder,
+  bool isScrollControlled = false,
+  Color bottomSheetColor,
```
I was concerned that `backgroundColor` might be ambiguous for the modal scrim.  Maybe that's being overly paranoid though, I'll make it backgrounColor.

dnfield committed:
Feedback

dnfield committed:
save before committing

dnfield committed:
Merge branch 'master' into bottom_sheet

dnfield committed:
Update scaffold.dart

rami-a commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -376,8 +437,14 @@ Future<T> showModalBottomSheet<T>({
 PersistentBottomSheetController<T> showBottomSheet<T>({
   @required BuildContext context,
   @required WidgetBuilder builder,
+  Color bottomSheetColor,
```
For consistency, can this be renamed to `backgroundColor` to match `showModalBottomSheet`

rami-a commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -1567,12 +1670,31 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
-  PersistentBottomSheetController<T> showBottomSheet<T>(WidgetBuilder builder) {
+  PersistentBottomSheetController<T> showBottomSheet<T>(
+    WidgetBuilder builder, {
+    Color bottomSheetColor,
```
For consistency, can this be renamed to `backgroundColor` to match `showModalBottomSheet`

dnfield replied:
Done

dnfield committed:
bottomSheetColor -> backgroundColor

rami-a commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -277,7 +277,7 @@ class _ModalBottomSheetState<T> extends State<_ModalBottomSheet<T>> {
                 delegate: _ModalBottomSheetLayout(animationValue, widget.isScrollControlled),
                 child: BottomSheet(
-                  color: widget.route.bottomSheetColor,
+                  color: widget.route.backgorundColor,
```
Nit-picking at this point, but typo for backgroundColor throughout this commit

dnfield replied:
Augh!  Will fix.  Thanks for catching that.

dnfield committed:
Typo

HansMuller commented on packages/flutter/test/material/modal_bottom_sheet_test.dart:
```diff
@@ -113,7 +113,7 @@ void main() {
     expect(find.text('BottomSheet'), findsOneWidget);
 
-    await tester.pump(); // bottom sheet dismiss animation starts
+     await tester.pump(); // bottom sheet dismiss animation starts
```
extra space here

HansMuller reviewed (approved):
LGTM with a few small potatoes comments 

Hooray!

dnfield committed:
DONE

dnfield committed:
last param

ThinkDigitalRepair commented:
Thanks, @dnfield for sticking with this! Are any of the issues mentioned in the original post or other comments still existing?

dnfield commented:
If there are, please file a bug.  This should be good though.

ThinkDigitalRepair commented:
Shouldn't this be DraggableScrollable**Bottom**Sheet? and shouldn't the name of the parameter be initialChildSizeRatio, not size? since 1 isn't a size, but a ratio. Same with the other parameters. Also, could we match the params of the other bottom sheets? Like the shape. Does this need to be a new issue?

dnfield commented:
This initially was implemented to support horizontal scrolling as well, but was removed due to complexity and lack of solid use cases.  It would be possible to reimplement

dnfield commented:
And in general it's probably a bad idea to specify absolute heights for this wide, since it will show up on different screens and form factors.

E.g. what if the phone is in landscape mode, or what if you're on a desktop or larger tablet

ThinkDigitalRepair commented:
Wouldn't that be our business to control? For example, I want a widget that works exactly like a bottomsheet that will cover most of my screen except the appbar.

dnfield commented:
You can just specify it as a percentage of the available space

ThinkDigitalRepair commented:
I think it's because I thought this was a bottomsheet
